datafusion/proto/src/physical_plan/mod.rs

```diff
@@ -394,17 +394,12 @@ impl AsExecutionPlan for PhysicalPlanNode {
                     vec![]
                 };
 
-                let input_schema = hash_agg
-                    .input_schema
-                    .as_ref()
-                    .ok_or_else(|| {
-                        DataFusionError::Internal(
-                            "input_schema in AggregateNode is missing.".to_owned(),
-                        )
-                    })?
-                    .clone();
-                let physical_schema: SchemaRef =
-                    SchemaRef::new((&input_schema).try_into()?);
+                let input_schema = hash_agg.input_schema.as_ref().ok_or_else(|| {
+                    DataFusionError::Internal(
+                        "input_schema in AggregateNode is missing.".to_owned(),
+                    )
+                })?;
+                let physical_schema: SchemaRef = SchemaRef::new(input_schema.try_into()?);
```

Comment:
The SchemaRef seems able create from ref of input_schema.

Reply:
I is confusing, but `SchemaRef` is an alias for `Arc<Schema>` which is fast and quick to copy

Reply:
The `input_schema` here is a `&protobuf::Schema`, and the `SchemaRef` is an alias of `Arc<physical_plan::Schema>`.

https://github.com/apache/arrow-datafusion/blob/0911f1523ec7088bae88684ecb9bca94aa553693/datafusion/proto/src/logical_plan/from_proto.rs#L372-L379

https://github.com/apache/arrow-datafusion/blob/0911f1523ec7088bae88684ecb9bca94aa553693/datafusion/proto/src/logical_plan/from_proto.rs#L609-L620

This really confusing me at first glance, it's consume a `&protobuf::Schema` when calling `try_into()`. Seems fields will be recreated anyway, I think the clone of `input_schema` could be eliminated.

I don't understand why these convert trait methods convert into `U` from a `&T` rather than a `T`, this makes implicit clones in convert methods.

https://github.com/apache/arrow-datafusion/blob/0911f1523ec7088bae88684ecb9bca94aa553693/datafusion/proto/src/physical_plan/mod.rs#L490

And I just notice this line could be replaced by `physical_schema`.

Reply:
I agree it is confusing. Any help to make it better would be most appreciated